builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+        pInactive = CreateInsertElement(pUndef, pInactive, static_cast<uint64_t>(0));
+        pInactive = CreateBitCast(pInactive, getInt32Ty());
+        Value* pResult = CreateSetInactive(pActive, pInactive);
+        pResult = CreateBitCast(pResult, pVectorType);
+        return CreateExtractElement(pResult, static_cast<uint64_t>(0));
+    }
+    else
+    {
+        Value* const pPreventControlFlow = CreateInlineAsmSideEffect(pActive);
+        return CreateIntrinsic(Intrinsic::amdgcn_set_inactive, pType, {pPreventControlFlow, pInactive});
+    }
+}
+
+// =====================================================================================================================
+// Create a call to mbcnt.
+Value* BuilderImplSubgroup::CreateMbcnt(
+    Value* const pValue) // [in] The value to pass to the mbcnt call, must be an i64.
+{
+    // Check that the type is definitely an i64.
+    LLPC_ASSERT(pValue->getType()->isIntegerTy(64));
+
+    // Cast i64 -> <2 x i32>.
+    Value* const pBitCast = CreateBitCast(pValue, VectorType::get(getInt32Ty(), 2));
+
+    Value* const pExtractLo = CreateExtractElement(pBitCast, getInt32(0));
+    Value* const pExtractHi = CreateExtractElement(pBitCast, getInt32(1));
+    CallInst* const pMbcntLo = CreateIntrinsic(Intrinsic::amdgcn_mbcnt_lo, {}, { pExtractLo, getInt32(0) });
+    return CreateIntrinsic(Intrinsic::amdgcn_mbcnt_hi, {}, { pExtractHi, pMbcntLo });
+}
+
+// =====================================================================================================================
+// Create a ds_swizzle bit mode pattern.
+uint16_t BuilderImplSubgroup::GetDSSwizzleBitMode(
+    uint8_t xorMask, // The xor mask (bits 10..14).
+    uint8_t orMask,  // The or mask (bits 5..9).
+    uint8_t andMask) // The and mask (bits 0..4).
+{
+    return (static_cast<uint16_t>(xorMask & 0x1f) << 10) |
+           (static_cast<uint16_t>(orMask & 0x1f) << 5) |
+           (andMask & 0x1f);
+}
+
+// =====================================================================================================================
+// Create a ds_swizzle quad mode pattern.
+uint16_t BuilderImplSubgroup::GetDSSwizzleQDMode(
+    uint8_t lane3, // The 3rd lane.
+    uint8_t lane2, // The 2nd lane.
+    uint8_t lane1, // The 1st lane.
+    uint8_t lane0) // The 0th lane.
+{
+    return 0x8000 | static_cast<uint16_t>((lane3 << 6) | ((lane2 & 0x3) << 4) | ((lane1 & 0x3) << 2) | ((lane0 & 0x3)));
+}
+
+// =====================================================================================================================
+// Create a map to i32 function. Many AMDGCN intrinsics only take i32's, so we need to massage input data into an i32
+// to allow us to call these intrinsics. This helper takes a function pointer, massage arguments, and passthrough
+// arguments and massages the mappedArgs into i32's before calling the function pointer. Note that all massage
+// arguments must have the same type.
+Value* BuilderImplSubgroup::CreateMapToI32(
```

Comment:
Please move it to base class BuilderImplBase, I suppose we need this function in multiple groups in the future.

Reply:
Good idea :)